cpp/include/cugraph/mtmg/resource_manager.hpp

```diff
@@ -0,0 +1,241 @@
+
+    int num_gpus_this_node;
+    RAFT_CUDA_TRY(cudaGetDeviceCount(&num_gpus_this_node));
+
+    CUGRAPH_EXPECTS((device_id.value() >= 0) && (device_id.value() < num_gpus_this_node),
+                    "device id out of range");
+
+    local_rank_map_.insert(std::pair(rank, device_id));
+
+    cudaSetDevice(device_id.value());
+
+    // FIXME: There is a bug in the cuda_memory_resource that results in a Hang.
+    //   using the pool resource as a work-around.
+    //
+    // Ultimately there should be some RMM parameters passed into this function
+    // (or the constructor of the object) to configure this behavior
+#if 0
+    auto per_device_it = per_device_rmm_resources_.insert(
+      std::pair{rank, std::make_shared<rmm::mr::cuda_memory_resource>()});
+#else
+    auto const [free, total] = rmm::detail::available_device_memory();
+    auto const min_alloc =
+      rmm::detail::align_down(std::min(free, total / 6), rmm::detail::CUDA_ALLOCATION_ALIGNMENT);
+
+    auto per_device_it = per_device_rmm_resources_.insert(
+      std::pair{rank,
+                rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
+                  std::make_shared<rmm::mr::cuda_memory_resource>(), min_alloc)});
+#endif
+
+    rmm::mr::set_per_device_resource(device_id, per_device_it.first->second.get());
+  }
+
+#if 0
+  /**
+   * @brief add a remote GPU to the resource manager.
+   *
+   * FIXME: Need some sort of comms information for the remote GPU here...
+   *
+   * @param rank    The rank to assign to the remote GPU
+   */
+  void register_remote_gpu(int rank, TBD const& remote_info);
+#endif
+
+  /**
+   * @brief Create an instance using a subset of the registered resources
+   *
+   * The selected set of resources will be configured as an instance manager.
+   * If @ranks_to_include is a proper subset of the registered resources,
+   * ranks will be renumbered into the range [0, @p ranks_to_use.size()), making
+   * it a proper configuration.
+   *
+   * @param ranks_to_use        a vector containing the ranks to include in the instance.
+   *   Must be a subset of the entire set of available ranks.
+   * @param instance_manager_id a ncclUniqueId that is shared by all processes participating
+   *   in this instance.  All processes must use the same ID in this call, it is up
+   *   to the calling code to share this ID properly before the call.
+   *
+   * @return unique pointer to instance manager
+   */
+  std::unique_ptr<instance_manager_t> create_instance_manager(
+    std::vector<int> ranks_to_include, ncclUniqueId instance_manager_id) const
+  {
+    std::for_each(
+      ranks_to_include.begin(), ranks_to_include.end(), [local_ranks = local_rank_map_](int rank) {
+        CUGRAPH_EXPECTS(local_ranks.find(rank) != local_ranks.end(),
+                        "requesting inclusion of an invalid rank");
+      });
+
+    std::vector<std::shared_ptr<ncclComm_t>> nccl_comms{};
+    std::vector<std::shared_ptr<raft::handle_t>> handles{};
```

Comment:
What's the rationale behind using shared_ptr here? Do we really expect multiple owner? Or instance manager will be the sole owner and all the other callers will just use it without managing the memory allocation/deallocation?